SpockOpaleye/src/ApiBase.hs

```diff
@@ -0,0 +1,47 @@
+{-# LANGUAGE FlexibleContexts       #-}
+{-# LANGUAGE FlexibleInstances      #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE MultiParamTypeClasses  #-}
+{-# LANGUAGE OverloadedStrings      #-}
+{-# LANGUAGE TemplateHaskell        #-}
+
+module ApiBase where
+
+import           Control.Lens
+import qualified Data.Profunctor.Product.Default as D
+import           Data.Time                       (UTCTime, getCurrentTime)
+import           Database.PostgreSQL.Simple
+import           DataTypes
+import           Opaleye
+import           OpaleyeDef
+import           OpaleyeTypes
+import           Prelude                         hiding (id)
+
+create_row ::(
+    HasCreatedat columnsW (Maybe (Column PGTimestamptz)),
+    HasUpdatedat columnsW (Column PGTimestamptz),
+    D.Default Constant incoming columnsW, D.Default QueryRunner returned row)
+    => Connection -> Table columnsW returned -> incoming -> IO row
+create_row conn table item = do
+  current_time <- fmap pgUTCTime getCurrentTime
+  let itemPg = (constant item) & createdat .~ (Just current_time) & updatedat .~ (current_time)
+  fmap head $ runInsertManyReturning conn table [itemPg] (\x -> x)
+
+update_row :: (
+    HasUpdatedat haskells UTCTime
+    , D.Default Constant haskells columnsW
+    , D.Default Constant item_id (Column PGInt4)
+    , HasId haskells item_id
+    , HasId columnsR (Column PGInt4)
+    )
+    => Connection -> Table columnsW columnsR -> item_id -> haskells -> IO haskells
+update_row conn table it_id item = do
```

Comment:
>The ID is the database identity of the record, which is what we're bothered with.

A database row can only have one state at a given time. If you want a in-memory representation of a database row, you have to make sure that there are no more than one instance of it at the same time in the scope. If you cannot gaurantee that, then you will have inconstant state in your app, when there are conflicting representations of the same database row, and upon saving to the db, you will lose data.

>In fact, if we want to be pedantic about it, I would NOT EXPORT the id setter lens at all. No one should be able to change the id of a record which represents a row in the DB. It has no practical purpose, and will result in bad things.

I am not sure preventing updates to id will make any difference. Because either way, you are not mutating in place. So there is really no difference between updating id, and updating all the fields while keeping the id same. Because they both are different instances.

There is no difference between the result of the following operations

Starting with two User records,

    u1 = User {id = 10, name = "max", age= 20}
    u2 = User {id = 15, name = "bob", age= 25}

Updating the id field of u1
    
    u3 = u1 & id .~ 15

is equalent to updating the name and age of u2 to that of u1, while keeping the id same
 
    u3 = u2 & name .~ "max" & age .~ 20

In both cases u3 has same values and is a different instance from either u3 or u2. 

>As soon as the DB moves out of the database and goes to any other "temporary holding area" (record in Haskell, data-structure in Redis, etc.) it is no longer an "authoritative representation" at all. 

It is still authoritative representation even after it leaves the DB. In fact, once it leaves the DB, the row in the DB cease to be the authoritative representation, because the data might be getting mutated as it passes through the app. The row in the DB won't be valid until the entity gets saved back to the database.

>And what exactly are we trying to prevent here? Can you give a real-life example of the use-case you are envisioning?

What we are trying to prevent is ending up with inconsistent data. I gave you an example earlier. You reply was 

>Whoa... stop right there. You don't get to "pick" which record should be saved to which ID! The ID is part of the record and self-identifies its row.

But that breaks down as soon as you have multiple instances of a record with the same ID.